### ISSUE TITLE
MIDI system common messages: Documentation doesn't match implementation

### ISSUE THREAD

**Issue by jamshark70** [supercollider/supercollider#5198]
## Environment

* SuperCollider version: 3.11 or before
* Operating system: Any

## Steps to reproduce

Using a Pd patch to generate messages (see [1]):

```supercollider
MIDIClient.init;

MIDIIn.connect(0, MIDIClient.sources.detect { |src| src.device.contains("Pure Data") }.uid);

MIDIdef.cc(\x, { |... args| args.debug("cc") });

MIDIdef.songPosition(\y, { |... args| args.debug("song pos") });

MIDIdef.songSelect(\y, { |... args| args.debug("song select") });

MIDIdef.tuneRequest(\y, { |... args| args.debug("tune req") });

MIDIdef.midiClock(\y, { |... args| args.debug("midiClock") });

MIDIdef.sysrt(\z, { |... args| args.debug("sysrt") });
```

Prints:

```
cc: [ 64, 1, 0, 8454145 ]  (this is only to be sure [midiout] is sending valid messages)
sysrt: [ 8192, 8454145, 0 ]
sysrt: [ 0, 8454145, 0 ]
sysrt: [ 0, 8454145, 6 ]
sysrt: [ 0, 8454145, 8 ]
midiClock: [ 8454145 ]
sysrt: [ 0, 8454145, 10 ]
sysrt: [ 0, 8454145, 11 ]
sysrt: [ 0, 8454145, 12 ]
sysrt: [ 0, 8454145, 14 ]
```

Corresponding to outgoing:

```
sending: symbol cc_for_testing
sending: symbol mtc_quarter_frame
sending: symbol song_position
sending: symbol song_select
sending: symbol tune_request
sending: symbol clock
sending: symbol start
sending: symbol continue
sending: symbol stop
sending: symbol active_sensing
```

## Expected vs. actual behavior

We carefully document `songPosition`, `songSelect`, `tuneRequest` and `midiClock` methods of MIDIFunc and MIDIdef.

Of these, only `MIDIdef.midiClock` functions as expected. `songPosition`, `songSelect` and `tuneRequest` responders are never called. They are forwarded to `sysrt` with different codes.

MIDIdef help does document:

> N.B. Because of SC's underlying low level MIDI implementation, there is no generic msgType and convenience method for System Common messages. Instead these are grouped with System Realtime under *sysrt below.

But then, immediately after stating that there are no convenience methods for system common messages, we go on to document several convenience methods (which do nothing!) -- this is misleading.

I can imagine two ways to proceed:

A. Remove both the documentation and implementation for the non-functioning convenience methods, and provide examples of how to receive these messages using `sysrt`.

B. Or, finish the implementation so that the message types work.

I'd suggest A -- these are not commonly used and I don't mind them being less convenient than MIDI note or CC messages.

Also, side note -- currently, both MTC quarter frame and song position messages arrive in `sysrt` with code = 0, so we have no way to differentiate them in SC.

[1] Pd patch
![pd-send-syscommon-midi](https://user-images.githubusercontent.com/318301/96201601-95a09080-0f8f-11eb-846b-17059ab12670.png)

**Comment by jamshark70** [comment]
I just realized that I made a really brain-dead copy/paste mistake in the test script -- most of the MIDIdefs have the same key! So of course they won't all register.

But, after fixing this, the result is the same, except that `tuneRequest` does work. Song position and some select still go through to sysrt.

```
MIDIClient.init;

MIDIIn.connect(0, MIDIClient.sources.detect { |src| src.device.contains("Pure Data") }.uid);

MIDIdef.cc(\x, { |... args| args.debug("cc") });

MIDIdef.songPosition(\a, { |... args| args.debug("song pos") });

MIDIdef.songSelect(\b, { |... args| args.debug("song select") });

MIDIdef.tuneRequest(\c, { |... args| args.debug("tune req") });

MIDIdef.midiClock(\d, { |... args| args.debug("midiClock") });

MIDIdef.sysrt(\z, { |... args| args.debug("sysrt") });
```

**Comment by muellmusik** [comment]
Hey James, I'm surprised as I thought all this worked. I certainly remember testing it, though I no longer have the max code – I think – I used for that. Apologies if this was mucked up somehow.

Can you test if trace works properly? Also maybe try a different source. 'both MTC quarter frame and song position messages arrive in sysrt with code = 0, so we have no way to differentiate them in SC' sounds wrong to me, but TBH I never use this stuff and it's been 10 years since I looked at it, so it's possible I'm wrong.

As far as fixes, I'd suggest B, as it seems like most of the work is also done. That is of course assuming that this is not a back end MIDI issue.

**Comment by muellmusik** [comment]
Okay, testing on Mac with Pocket MIDI. Song Select, Song Position, MIDI Clock, Active Sense, and Tune Request all fine. I tested tick from PD using a super simplified version of your patch and that worked too. The larger messages didn't work from PD though, and weren't received in either SC or Pocket MIDI.

In fairness the documentation for midiout in PD is crap, so I have no idea if it's doing what we seem to think it is. But my guess is the issue is either the patch, PD itself, or SC's MIDI back end on linux. On Mac this seems all completely fine.

**Comment by muellmusik** [comment]
Well, not completely I suppose: The nota bene looks vestigial from before this stuff was added and should probably be removed. :-)

**Comment by jamshark70** [comment]
Since you have a reliable source for these MIDI messages, would you mind using pd or max to dump, byte for byte, the exact correct format for all of these message types?

To create the patch, I used two approaches:

- Read the MIDI spec and build messages by hand -- e.g., https://www.midi.org/specifications-old/item/table-1-summary-of-midi-message says song position pointer is 2r11110010 = 242 followed by 2 data bytes. So I wrote "242 101 2."

- Use SC's MIDIOut to send example messages to Pd's [midiin] and [midirealtimein], and copy the bytes.

For this to fail, both my reading of the MIDI spec would have to be wrong *and* MIDIOut would also have to be wrong. That's not impossible, but seems unlikely.

I'm thoroughly, entirely confused about the dispatch mechanism. What I expected to find in SC_CoreMIDI.cpp (looking here first, since it seems that everything is fine and Mac) is that song position, song select, tune request and clock messages would dispatch to separate methods (to go through to the different message types). Instead, they all go through `doSysrtAction`. So, there must be something in the class library to dispatch these to different locations. I've searched ResponseDefs.sc for a good half an hour now, and I just don't see it. There is `sysIndices` but it's used only for tracing. I guess there's something about the `index` in some of the dispatchers, but (maybe it's just me) I find this code style to be nearly impenetrable. Lacking comments or documentation, I'm at a loss (and you know that I'm generally pretty good at reading code).

So at this point, I can't start to guess how to fix it in Linux because I can't yet understand how it works in the first place.

**Comment by jamshark70** [comment]
A stab in half-darkness -- one problem in SC_AlsaMIDI.cpp (maybe not the only problem, but one problem) is that the song position etc. messages seem to assume that `evt->data.control.channel` contains a value that differentiates the message types, but in fact, this seems to be 0 for all of these message types.

The ones that are broken begin like this:

```
        case SND_SEQ_EVENT_SONGSEL: // song sel
            ++g->sp;
            SetInt(g->sp, evt->data.control.channel);
```

The ones that are not broken (e.g. tune request) begin like this:

```
        case SND_SEQ_EVENT_TUNE_REQUEST: // tuning request
            ++g->sp;
            SetInt(g->sp, 0x6);
```

But this doesn't completely fix it. Compare these:

```
        case SND_SEQ_EVENT_CLOCK: // clock
            ++g->sp;
            SetInt(g->sp, 0x8);
            ++g->sp;
            SetInt(g->sp, 0);
            runInterpreter(g, s_midiSysrtAction, 4);
            break;
        case SND_SEQ_EVENT_START: // start
            ++g->sp;
            SetInt(g->sp, 0xA);
            ++g->sp;
            SetInt(g->sp, 0);
            runInterpreter(g, s_midiSysrtAction, 4);
            break;
```

Identical in format. The first one (0x8) dispatches to a `midiClock` responder. The second one should dispatch to a `start` responder, but it doesn't. I can't explain that at the moment.

**Comment by muellmusik** [comment]
> Since you have a reliable source for these MIDI messages, would you mind using pd or max to dump, byte for byte, the exact correct format for all of these message types?

Active sense - FE
MIDI Clock - F8
Song Select - F3 00 (for song 1)
Song Position - F2 00 00 (for position 0)
Stop - FC
Continue - FB
Start - FA
Tune Request - F6
Tick - F9

Mostly checked with Pocket MIDI which is available for Windows, btw.

**Comment by muellmusik** [comment]
> I'm thoroughly, entirely confused about the dispatch mechanism.

MIDIFuncs/defs have a dispatcher object (generally the default dispatcher object for that type). It registers at the low level MIDIIn callback for that message type. You are correct that there is a single sysrt callback in MIDIIn. (I don't recall offhand why that is that way, but my guess would be that it preexisted the message specific MIDIFunc implementations and it would seem pointless to hit two low level callbacks for every sysrt message.)

The dispatchers for sysrt hold the MIDIFunc's action funcs in FunctionLists within an 'active' dict by message number.

callback->dispatcher->activeFuncsForSysRtMessageNum

It's actually quite simple at dispatch stage. It's true there is some complexity in how it's set up. What it also is is _very_ fast. We did a lot of benchmarking when we did this, and it really minimises tests at every stage. By encapsulating dispatch into another object, it also makes that modular and customisable, which was a request at the time.

In any case, as I said, it works on Mac, so I think the problem is the back end. I would suggest checking at low level MIDIIn that sysrt callbacks are passing the correct values.

**Comment by muellmusik** [comment]
> A stab in half-darkness -- one problem in SC_AlsaMIDI.cpp

Hmm. I confess I know next to nothing about the AlsaMIDI implementation (or AlsaMIDI in general) and currently have nothing to test it on. It's possible of course that the low-level sysrt callback never worked on Linux or Windows, and that this issue predates MIDIFunc (or even the older responders). I thought Marije tested all this on Linux at the time, but it was something like 10 years ago, so my memory may be faulty. :-(

**Comment by muellmusik** [comment]
> messages seem to assume that `evt->data.control.channel` contains a value that differentiates the message types, but in fact, this seems to be 0 for all of these message types.

Ah hopefully an easy fix then. Since it's already detected message type presumably the correct approach is to pass the expected value?

**Comment by jamshark70** [comment]
> Since it's already detected message type presumably the correct approach is to pass the expected value?

OK, I've added a PR. It does pass my test case.

> Mostly checked with Pocket MIDI which is available for Windows, btw.

I am sending equivalently formatted messages (though I'm using non-zero data bytes for song select and position -- this revealed a different bug in decoding the song position value, which I also fixed in the PR). And the test messages from Pd are received correctly in Linux.

If there is a [midiout] bug in Pd, then it seems to be platform-specific for Mac. (Or perhaps Mac's IAC bus is doing something funny.)

> MIDIFuncs/defs have a dispatcher object (generally the default dispatcher object for that type). It registers at the low level MIDIIn callback for that message type.

This might be a topic to discuss on the mailing list or forum rather than here.

I don't have any complaints about the object structure -- it's a good design and I do see the performance benefits. My experience with it has been that, when something goes wrong and it's necessary to trace through the object structure and data flow, this part of the class library is considerably more difficult than other areas.

It's a boring and labor-intensive documentation task, but it might help to have a table of which dispatchers are used for which scenarios. (I did eventually figure out that the msgNum in the dispatcher determines which functions fire for the incoming sysrt messages -- but it took a much longer time than I expected. For the sake of future maintenance, we really should write some of this down so that others later on can get to the answer more quickly -- otherwise, people might simply give up troubleshooting and then issues don't get fixed.)
